### PR TITLE
Fix bugprone-use-after-move clang-tidy warnings

### PR DESCRIPTION
**Pull request by SiarheiFedartsou**
Signed-off-by: Siarhei Fedartsou <siarhei.fedartsou@gmail.com>

### CODE REVIEW

File: generator/camera_info_collector.cpp [hunk 1]
```diff
@@ -137,6 +136,7 @@ bool CamerasInfoCollector::ParseIntermediateInfo(
                    "lat(", lat, "), lon(", lon, ")"));
     }
 
+    std::vector<routing::SpeedCameraMwmPosition> ways;
```
Comment (SiarheiFedartsou) [inline generator/camera_info_collector.cpp:139]: I am not sure what was intention here initially, so fixed it with my understanding of surrounding code, but please double check it.

Comment (SiarheiFedartsou) [reply generator/camera_info_collector.cpp:139]: Just to better understand: `ways` are declared outside of loop and being moved inside this loop.

File: drape_frontend/drape_engine.cpp [hunk 2]
```diff
@@ -92,8 +92,6 @@ DrapeEngine::DrapeEngine(Params && params)
                                     std::move(effects),
                                     params.m_onGraphicsContextInitialized);
 
-  m_frontend = make_unique_dp<FrontendRenderer>(std::move(frParams));
```
[the remark this answers is not in the document]
Comment (biodranik) [reply drape_frontend/drape_engine.cpp:95]: Wow wow wow! @vng look, we have a UB here!

File: generator/camera_info_collector.cpp (hunk 1 above)
Comment (biodranik) [reply generator/camera_info_collector.cpp:139]: Good catch. But the better fix would be to call `ways.clear()` to explicitly set the empty state, and to avoid unnecessary allocations in each loop iteration.

File: generator/osm_source.cpp [hunk 3]
```diff
@@ -120,7 +120,7 @@ void ProcessOsmElementsFromXML(SourceReader & stream, std::function<void(OsmElem
   while (processorOsmElementsFromXml.TryRead(element))
   {
     processor(std::move(element));
-    element.Clear();
+    element = {};
```
Comment (biodranik) [inline generator/osm_source.cpp:123]: What was wrong with Clear?

File: generator/world_roads_builder/world_roads_builder.cpp [hunk 4]
```diff
@@ -69,8 +69,10 @@ RoadsFromOsm GetRoadsFromOsm(generator::SourceReader & reader,
     {
       std::string const & highway = e.GetTag("highway");
 
-      if (!highway.empty() && base::IsExist(highways, highway))
-        roadsFromOsm.m_ways[highway].emplace(e.m_id, RoadData({}, std::move(e)));
+      if (!highway.empty() && base::IsExist(highways, highway)) {
+        auto id = e.m_id;
```
Comment (biodranik) [inline generator/world_roads_builder/world_roads_builder.cpp:73]:
```suggestion
      if (!highway.empty() && base::IsExist(highways, highway))
      {
        auto id = e.m_id;
```

File: generator/camera_info_collector.cpp (hunk 1 above)
Comment (vng) [reply generator/camera_info_collector.cpp:139]: Well, I'd prefer to declare ways vector as close to the using scope as possible.

File: generator/osm_source.cpp (hunk 3 above)
Comment (vng) [reply generator/osm_source.cpp:123]: Hm, need to check what Clear actually does, but empty initialization is definitely better than some Clear after move.

File: generator/camera_info_collector.cpp (hunk 1 above)
Comment (biodranik) [reply generator/camera_info_collector.cpp:139]: Then buffer_vector can be a better candidate. Avoiding unnecessary allocations is a very good pattern.

Comment (vng) [reply generator/camera_info_collector.cpp:139]: This is generator, so no need to check anything here. And buffer_vector also should treat carefully (it sucks sometimes) :)

File: generator/osm_source.cpp (hunk 3 above)
Comment (biodranik) [reply generator/osm_source.cpp:123]: Why is it _definitely_ better? What are the pros and cons?

File: generator/camera_info_collector.cpp (hunk 1 above)
Comment (biodranik) [reply generator/camera_info_collector.cpp:139]: _Faster_ generator is also a great thing )

File: generator/osm_source.cpp (hunk 3 above)
Comment (vng) [reply generator/osm_source.cpp:123]: It strongly depends on what Clear actually does. Clear call can _depend_ on some inner state, which is undefined after move. Clear call can reset/initialize not all class inner fields.

Comment (biodranik) [reply generator/osm_source.cpp:123]: Then _Clear_ should either be _fixed_ to provide expected behavior or it should be _removed_ from the class.

Comment (vng) [reply generator/osm_source.cpp:123]: Hm, ok. Then write comment here.

Comment (SiarheiFedartsou) [reply generator/osm_source.cpp:123]: > 
element.Clear(); should run faster because the container capacity does not change and we reduce a number of memory reallocations in the future. I would ask for performance tests. (The secret is that it is important at this place.)

Which containers do you mean here? These ones? https://github.com/organicmaps/organicmaps/blob/82f71e159ea5198ce3a82471a7129f41e816e6e2/generator/osm_element.hpp#L167

I am not C++ expert, but it seems that after moving they become invalid and we shouldn't call `clear()` on them here https://github.com/organicmaps/organicmaps/blob/82f71e159ea5198ce3a82471a7129f41e816e6e2/generator/osm_element.hpp#L91

Should we? 

In other words:
1. We move `OsmElement` in this place where I fixed it (i.e. its internal containers are in invalid state now)
2. Then call `OsmElement::Clear`, which in its turn calls `clear()` on `std::vector/std::string` which are now in invalid state after moving

WDYT?

Comment (vng) [reply generator/osm_source.cpp:123]: BTW, not sure (probably it's not specified) but vector = {} should keep the capacity.

Comment (vng) [reply generator/osm_source.cpp:123]: Moreover, if we move the vector, there is no sense think about capacity of previous vector. The _capacity_ buffer definitely will be also moved.

Comment (biodranik) [reply generator/osm_source.cpp:123]: @SiarheiFedartsou the object after move is not _invalid_. It is `valid but in unspecified state`. So calling `clear()` is perfectly fine, while calling pop_back() is UB.

https://stackoverflow.com/a/9168917

Comment (SiarheiFedartsou) [reply generator/osm_source.cpp:123]: hm, interesting, didn't know that. BUT I doubt it somehow helps with allocations(I assume `clear` will allocate new buffer in this case) + if in the case like this we can avoid "use after move", why don't we do that to avoid bugs in the future? after all who knows how we may want to modify this `Clear` method in the future?

Comment (vng) [reply generator/osm_source.cpp:123]: Well again, = {} is better and more _code-clear_ here than "Clear".

Comment (maksimandrianov) [reply generator/osm_source.cpp:123]: When I wrote this, I also knew about  = {}. I did performance measurements, you can try it too. The choice is yours:)

Comment (biodranik) [reply generator/osm_source.cpp:123]: Benchmark shows that Clear is faster: https://godbolt.org/z/3TE7WPT4h

I vote for the faster code.

Comment (SiarheiFedartsou) [reply generator/osm_source.cpp:123]: @biodranik vector is not moved before `clear` in this benchmark - I bet it should be no difference if you start moving it (at least I cannot imagine what magic clear can do to be faster on moved vector - memory is already moved and it has to allocate new buffer).

Unfortunately I am not near my laptop right now, but will try to “play” with this benchmark later.

File: generator/camera_info_collector.cpp (hunk 1 above)
Comment (vng) [reply generator/camera_info_collector.cpp:139]: You won't get anything faster. Code readability is much better here.

File: generator/osm_source.cpp (hunk 3 above)
Comment (vng) [reply generator/osm_source.cpp:123]: Hm, I also can't realize what can be "slow" here. I suspect that correct (with move) benchmark will have bigger threshold error than actual difference between clear or default initialization.

Comment (vng) [reply generator/osm_source.cpp:123]: But I don't mind keep Clear here, write comment and merge this PR at last.

File: generator/camera_info_collector.cpp (hunk 1 above)
Comment (biodranik) [reply generator/camera_info_collector.cpp:139]: Added move for the honest comparison. Clear is 3x-5x times faster than reassign. I'm still strongly voting for the faster code. Clear is very _clear_ for readability.

https://godbolt.org/z/9hsWdeKGr

Comment (vng) [reply generator/camera_info_collector.cpp:139]: I don't mind. But looks like benchmark was in debug :) Here is my results:
Screenshot 2022-12-17 at 18 56 25

The actual dispersion is bigger than actual timings ..

Comment (biodranik) [reply generator/camera_info_collector.cpp:139]: You're right, in the second link the -O3 switch was reset. You need to add -std=c++17 btw for the correct code generation.
Then results are not as straight, and depend on the compiler (clang is better on linux than gcc):
```
Clear 0
Clear 0
Clear 41
Clear 83
Clear 0
Clear 0
Reassign 209
Reassign 167
Reassign 41
Reassign 41
Reassign 42
Reassign 42
```

File: generator/osm_source.cpp (hunk 3 above)
Comment (biodranik) [inline generator/osm_source.cpp:123]:
```suggestion
    element.Clear();
```

File: generator/osm_source.cpp [hunk 5]
```diff
@@ -144,7 +144,7 @@ void ProcessOsmElementsFromO5M(SourceReader & stream, std::function<void(OsmElem
   while (processorOsmElementsFromO5M.TryRead(element))
   {
     processor(std::move(element));
-    element.Clear();
+    element = {};
```
Comment (biodranik) [inline generator/osm_source.cpp:147]:
```suggestion
    element.Clear();
```

@SiarheiFedartsou what was clang-tidy complaint about Clear? Why did you decide to change it with = {}?

Comment (SiarheiFedartsou) [reply generator/osm_source.cpp:147]: It was used after move(one line above) - in vast majority of cases it is a bug

Comment (vng) [reply generator/osm_source.cpp:147]: Could you please update this PR?
- put Clear() and write comment
- use normal { style (from new line) instead of KR-style

We will merge it and build beta version today.

Comment (biodranik) [reply generator/osm_source.cpp:147]: Here it is not a bug, right? Let's fix and move on. Other fixes here are important for the production release.

**Comment by SiarheiFedartsou** [comment]
@biodranik @vng 
I updated PR, ptal again

File: generator/osm_source.cpp [hunk 6]
```diff
@@ -120,6 +120,7 @@ void ProcessOsmElementsFromXML(SourceReader & stream, std::function<void(OsmElem
   while (processorOsmElementsFromXml.TryRead(element))
   {
     processor(std::move(element));
+    // it is safe to use `element` here as `Clear` will restore state after move
```
Comment (biodranik) [inline generator/osm_source.cpp:123]:
```suggestion
    // It is safe to use `element` here as `Clear` will restore the state after the move.
```

File: generator/osm_source.cpp [hunk 7]
```diff
@@ -144,6 +145,7 @@ void ProcessOsmElementsFromO5M(SourceReader & stream, std::function<void(OsmElem
   while (processorOsmElementsFromO5M.TryRead(element))
   {
     processor(std::move(element));
+    // it is safe to use `element` here as `Clear` will restore state after move
```
Comment (biodranik) [inline generator/osm_source.cpp:148]:
```suggestion
    // It is safe to use `element` here as `Clear` will restore the state after the move.
```

**Review by biodranik — APPROVED** [review]

**Review by biodranik — APPROVED** [review]
@SiarheiFedartsou thank you very much for the fast reaction and for the fix!
